Sources/LeafKit/LeafSyntax/LeafSyntax.swift

```diff
@@ -18,29 +18,133 @@ public enum ConditionalSyntax {
-                        // export exists, inject body
-                        extended += export.body
-                    } else {
-                        // any unsatisfied import will continue
-                        // and can be satisfied later
-                        extended.append(syntax)
-                    }
-                default:
-                    extended.append(syntax)
-            }
 
+// temporary addition
+extension Syntax: BodiedSyntax  {
+    func externals() -> Set<String> {
+        switch self {
+            case .conditional(let bS as BodiedSyntax),
+                 .custom(let bS as BodiedSyntax),
+                 .export(let bS as BodiedSyntax),
+                 .extend(let bS as BodiedSyntax),
+                 .loop(let bS as BodiedSyntax): return bS.externals()
+            default: return .init()
+        }
+    }
+    
+    func imports() -> Set<String> {
+        switch self {
+            case .import(let i): return .init(arrayLiteral: i.key)
+            case .conditional(let bS as BodiedSyntax),
+                 .custom(let bS as BodiedSyntax),
+                 .export(let bS as BodiedSyntax),
+                 .extend(let bS as BodiedSyntax),
+                 .expression(let bS as BodiedSyntax),
+                 .loop(let bS as BodiedSyntax): return bS.imports()
+            // .variable, .raw
+            default: return .init()
+        }
+    }
+    
+    func inlineRefs(_ externals: [String: LeafAST], _ imports: [String: Export]) -> [Syntax] {
+        var result = [Syntax]()
+        switch self {
+            case .import(let im):
+                let ast = imports[im.key]?.body
+                if let ast = ast {
+                    // If an export exists for this import, inline it
+                    ast.forEach { result += $0.inlineRefs(externals, imports) }
+                } else {
+                    // Otherwise just keep itself
+                    result.append(self)
+                }
+            // Recursively inline single Syntaxes
+            case .conditional(let bS as BodiedSyntax),
+                 .custom(let bS as BodiedSyntax),
+                 .export(let bS as BodiedSyntax),
+                 .extend(let bS as BodiedSyntax),
+                 .loop(let bS as BodiedSyntax): result += bS.inlineRefs(externals, imports)
+            case .expression(let pDA): result.append(.expression(pDA.inlineImports(imports)))
+            // .variable, .raw
+            default: result.append(self)
         }
-        return extended
+        return result
+    }
+}
+
+internal protocol BodiedSyntax {
+    func externals() -> Set<String>
+    func imports() -> Set<String>
+    func inlineRefs(_ externals: [String: LeafAST], _ imports: [String: Syntax.Export]) -> [Syntax]
+}
+
+extension Array: BodiedSyntax where Element == Syntax {
+    func externals() -> Set<String> {
+        var result = Set<String>()
+        _ = self.map { result.formUnion( $0.externals()) }
```

Comment:
Similar to above, this could be one of:
- `return reduce(Set<String>()) { $0.formUnion($1.externals()) }`
- `return Set(lazy.flatMap { $0.externals() })`